Use govuk-frontend v5 [DO NOT MERGE]

patrickpatrickpatrick opened the pull request:
[Trello](https://trello.com/c/6LsKqHIO/2231-update-version-of-govuk-frontend-in-govukpublishingcomponents-l)

## What

Make changes in govuk_publishing_components to allow us to update to govuk-frontend v5.

### Asset Changes

- Update references to assets (config, Javascript, CSS, fonts, images) in govuk-frontend
- Remove references to the crown fallback image
- Remove references to 
  - `assets/images/govuk-apple-touch-icon.png`
  - `assets/images/govuk-apple-touch-icon-152x152.png`
  - `assets/images/govuk-apple-touch-icon-167x167.png`
  - `assets/images/govuk-apple-touch-icon-180x180.png`
  - `assets/images/govuk-mask-icon.svg`
- Add references to
  - `assets/manifest.json`
  - `assets/images/govuk-icon-180.png`
  - `assets/images/govuk-icon-mask.svg`
- Change logo in-line svg to be the new svg that includes GOV.UK
  - Rework how the logo is included in `layout_super_navigation_header`

### Javascript Changes

- Add necessary script to append new required class to body for govuk-frontend
- Update initialisation method to be the same as the new govuk-frontend approach for any JS module that uses `GOVUK.Modules`
- Update tests to use new initialisation method for any of the JS modules I had to update
- Remove import of `details.js` from govuk-frontend as the `details` component no longer uses Javascript
- Use [Terser](https://github.com/ahorek/terser-ruby) instead of [Uglifier](https://github.com/lautis/uglifier) to compress Javascript

### Stylesheet Changes
- Remove `$legacy` from the `govuk-colour` mixin (this has been removed from govuk-frontend)
- Remove `$govuk-new-link-styles` feature flag (new link styles now default, feature flag removed)

### Component specific changes

- Components using `tag` component now `titleize` the tag text to adopt the new tag design
   * Update the `header_logo` and `phase_banner` components
   * Update tests of `layout_header` and `phase_banner`
   * Tweak text colour of `gem-c-environment-tag` to ensure contrast (default tag colour is now black).
- `warning_text` component no longer uses the `govuk-warning-text__assistive` class
  * Replaced all instances of `govuk-warning-text__assistive` with `govuk-visually-hidden` as specified in the release notes
  * Update tests of warning_text component

### Config changes
- Add new configuration option `use_es6_components`
  * If true then `layout_for_public` will try to load the ES6 (govuk-frontend) using components from the file `es6-components.js`

### Doc updates
- Add suggested way to include JS in Rails applications in light of the browser target change that has occurred in govuk-frontend v5.

## What I Haven't Changed

This PR is for things that **have** to change. To keep the PR size as manageable as possible (sorry, it is still very long) there are pieces of work that I think should be separate PRs:

- If a module isn't using the `GOVUK.Modules` object (i.e. most of the analytics modules and most of the Ga4 modules) then it haven't been updated to use the new initialisation method introduced by govuk-frontend v5. 
- Opening the component guide in browsers that don't support `type="module"` will result in some ungraceful erroring. This is because in this PR there isn't a way to individually load the JS that contains ES6 code. In frontend rendering applications I have established a pattern to follow to do this and it could be implemented in them gem also (but in a separate PR). 

## Why

There are several breaking changes in the latest version of gouk-frontend which requires us to update govuk_publishing_components. [Please read the full release notes](https://github.com/alphagov/govuk-frontend/releases/tag/v5.0.0-beta.2). The changes that are specifically being addressed here are:

- Update package file paths
- Remove Internet Explorer 8 stylesheets, settings and mixins
- Remove the fallback GOV.UK crown logo from your HTML
- Removal of Javascript polyfills for Internet Explorer (IE) 8-11
- Update the <script> snippet at the top of your <body> tag
- Check your tags align with design changes to the Tag component
- New link styles are now enabled by default
- Update the HTML for warning text


## Visual Differences

Percy: https://percy.io/55ba5e1a/govuk_publishing_components/builds/33548508/changed/1838427854?browser=firefox&browser_ids=57&subcategories=unreviewed%2Cchanges_requested&viewLayout=overlay&viewMode=new&width=1280&widths=375%2C1280

### Layout Header

The reason for the visual change here is that `.gem-c-header__logotype` was causing the line-height to increase, now this class is removed, the total height of the layout header has decreased from 72.66px to exactly 70px.

#### Before

<img width="951" alt="layout-header-before" src="https://github.com/alphagov/govuk_publishing_components/assets/28779939/201b7b02-7672-4fa2-a799-de5689bb2bfd">
<img width="962" alt="layout-header-before-sizes" src="https://github.com/alphagov/govuk_publishing_components/assets/28779939/c7e005b3-f3c4-4777-9098-ea9f97440dca">

#### After

<img width="952" alt="layout-header-after" src="https://github.com/alphagov/govuk_publishing_components/assets/28779939/136b8a91-8e50-42fc-9521-5b361fd423cc">
<img width="952" alt="layout-header-after-sizes" src="https://github.com/alphagov/govuk_publishing_components/assets/28779939/bcff5869-e1ed-4dbe-b770-ea3063cb8af9">

### Public Layout

Percy is also showing a visual change in public layout wherever the layout header component is used, this is to be expected following the change mentioned above.

| Before | After |
| --- | --- |
| ![image](https://github.com/alphagov/govuk_publishing_components/assets/3727504/a0059c14-e174-4b85-891a-6d4fa327a6f3) |  ![image](https://github.com/alphagov/govuk_publishing_components/assets/3727504/28ca23f7-eec1-48fc-b606-2bbff0da3e0e) |

### Phase Banner

The phase banner will now use the new design included in v5 of the govuk-frontend.

| Before | After |
| --- | --- |
| ![68e7957767c852045d2522b12d04913705f749b970d165415e8d84fe60f1726b](https://github.com/alphagov/govuk_publishing_components/assets/3727504/60574979-03b4-4402-991c-2f18a9339baa) |  ![b3460582d31935951cd58afacd27118567d2a311fabd59ba12e8f2a5ab5a430d](https://github.com/alphagov/govuk_publishing_components/assets/3727504/97656376-a3d2-4c84-a86a-ea11b0d841b6) |


### Form inputs

The visual change show in Percy for the form input component is coming from a fix in version 5.1.0 of govuk-frontend - https://github.com/alphagov/govuk-frontend/blob/main/CHANGELOG.md#510-feature-release

| Before | After |
| --- | --- |
| <img width="275" alt="form-input-prefix-before" src="https://github.com/alphagov/govuk_publishing_components/assets/28779939/47c844fe-3009-40ea-93aa-d2e454663bb5"> | <img width="298" alt="form-input-prefix-after" src="https://github.com/alphagov/govuk_publishing_components/assets/28779939/251a7b49-5975-4c94-ba75-83c50811d2b5"> |

patrickpatrickpatrick committed:
Update govuk-frontend to v5.1.0

patrickpatrickpatrick committed:
Use govuk-frontend v5 component initialisation

In govuk-frontend v5, initialisation has been moved to creation of
component JS module. This means that `init()` does not need to be
called to initialise a component, it is instead called in the
constructor.

To ensure that modules from govuk-frontend and modules from the gem
can be initialised easily in the same `initAll()` method, the modules
of the gem have been changed to use the new initialisation method. This
means moving `init()` to the constructor of each module. It also means
updating the `initAll()` method to not call 'init()' of each module.

patrickpatrickpatrick committed:
Update govuk-frontend asset paths

In govuk-frontend v5 the asset path has changed from:

`govuk-frontend/govuk/`

to

`govuk-frontend/govuk/dist`

so we have to change references to the old path.

patrickpatrickpatrick committed:
Remove $legacy from govuk-colour

govuk-frontend v5 deprecates the `$legacy` param in the `govuk-colour`
mixin. Continuing to use it produces warnings for each instance of it
in the CSS. It can be safely removed as alphagov/govuk-frontend#3576
notes that `it is non-operational and the parameter is only maintained to
prevent compilation errors`.

patrickpatrickpatrick committed:
Update tests to use new initialisation method

The modules in govuk_publishing_components have been updated to use the
new method of initialisation. To get tests to work correctly, we also
need to update how the modules are initialised in the tests themselves
as they are individually initialised instead of using `initAll()`

domoscargin commented on app/assets/javascripts/govuk_publishing_components/components/accordion.js:
```diff
@@ -1,10 +1,11 @@
 window.GOVUK = window.GOVUK || {}
 window.GOVUK.Modules = window.GOVUK.Modules || {}
-window.GOVUK.Modules.GovukAccordion = window.GOVUKFrontend.Accordion;
+window.GOVUK.Modules.GovukAccordion = window.GOVUKFrontend.Accordion.Accordion;
```
:bug: Hi @patrickpatrickpatrick - you've helped us uncover a bug here with our named exports. We'll hopefully patch this shortly to get rid of this odd duplication (so this import will just be `window.GOVUKFrotnend.Accordion`).

patrickpatrickpatrick replied:
Ah great, that makes sense!

colinrotherham replied:
Thanks again for this

We hadn't realised that Rollup handles naming for us now

Fixed in https://github.com/alphagov/govuk-frontend/pull/4444 and will be in the next beta

colinrotherham replied:
@patrickpatrickpatrick Released today in `beta.2`

* https://github.com/alphagov/govuk-frontend/releases/tag/v5.0.0-beta.2

patrickpatrickpatrick replied:
Thanks, I have updated the code :)

patrickpatrickpatrick committed:
Remove references to the crown fallback

In govuk-frontend v5.0.0, the crown fallback image for IE8 has been
removed so we need to remove all references to this asset in the gem.

patrickpatrickpatrick committed:
Remove references to govuk-frontend JS polyfills

In govuk-frontend v5.0.0, JS polyfills have been removed so therefore we
need to remove all references to these in the gem.

patrickpatrickpatrick committed:
Update tag component in light of govuk-frontend v5

govuk-frontend v5 introduces some changes to the tag component. The one
that has the most impact, is that the tag is now designed not to use
ALL CAPS text anymore. Instead it is to be titlecase. In addition there
are new background colours introduced which means that the styling for
some tags in govuk_publishing_components has to be changed to maintain
sufficient contrast. Tests have been updated to anticipate titlecase
tag text instead of ALL CAPS text.

patrickpatrickpatrick committed:
Remove `$govuk-new-link-styles` feature flag

In govuk-frontend v5, the new link styles are default and this feature
flag has been removed.

patrickpatrickpatrick committed:
Replace `govuk-warning-text__assistive` class

In govuk-frontend v5, the `govuk-warning-text__assistive` has been
removed. Instead `govuk-visually-hidden` is to be used. Tests that
look for `govuk-warning-text__assistive` have been updated.

patrickpatrickpatrick committed:
Use Terser instead of Uglifier

govuk-frontend v5 now targets browsers that support ES6.

This means that the UMD modules used in govuk_publishing_components from govuk-frontend
use features of ES6 and so it means that Uglifier can't be used anymore because it only supports ES5.

As well as installing Terser and updating the config, this commit also contains a patch for getting Terser working. Sprockets doesn't have an built-in loader for Terser so we need to add this functionality.

patrickpatrickpatrick commented:
The changes made here will mean that the component guide will no longer use Javascript correctly for certain components when opened in Internet Explorer 11. Currently this will results in errors in the console. Although this isn't ideal, personally I think this could be sorted out in separate branch? I'd assume the component guide is mostly used by developers and those developers won't be using Internet Explorer 11. It should be fixed but I don't think it's a pressing issue.

andysellick commented:
> The changes made here will mean that the component guide will no longer use Javascript correctly for certain components when opened in Internet Explorer 11. Currently this will results in errors in the console (in IE 11).

Interesting - can you explain more about why this will cause console errors? I thought the introduction of JS modules meant that it would fail silently for IE11. Is there something different about the component guide that causes this?

patrickpatrickpatrick commented:
> > The changes made here will mean that the component guide will no longer use Javascript correctly for certain components when opened in Internet Explorer 11. Currently this will results in errors in the console (in IE 11).
> 
> Interesting - can you explain more about why this will cause console errors? I thought the introduction of JS modules meant that it would fail silently for IE11. Is there something different about the component guide that causes this?

The issue is the scripts being evaluated at all by IE11. The modules need to be wrapped in a `type="module"` tag for IE11 to 'skip' over them. I have actually started to implement how we would do this in a PRs for [static](https://github.com/alphagov/static/pull/3182) and [collections](https://github.com/alphagov/collections/pull/3436) and it wouldn't be difficult to implement. It's just if it should be included in this initial PR or should be added in a seperate one (to keep this one as manageable as possible).

andysellick commented:
> It's just if it should be included in this initial PR or should be added in a separate one (to keep this one as manageable as possible as it is already quite long).

Thanks for the explanation, very helpful. Sounds sensible to put that fix in a separate PR, agreed. Shout if you want any help!

colinrotherham commented:
@patrickpatrickpatrick Probably worth a mention, but in readiness of `type="module"` we had some great feedback from our beta releases in https://github.com/alphagov/govuk-frontend/discussions/4389#discussioncomment-7402623 to look at

Reminders that JavaScript code in `<script type="module">`

* runs in ["strict mode"](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Strict_mode) but without `"use strict"`
* runs before [DOMContentLoaded](https://developer.mozilla.org/en-US/docs/Web/API/Document/DOMContentLoaded_event) like [`<script defer>`](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/script#defer)

The script contents itself doesn't need ES module exports, but good to catch any issues early such as:

>Strict mode applies to entire scripts or to individual functions. It doesn't apply to [block statements](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Statements/block) enclosed in {} braces; attempting to apply it to such contexts does nothing. [eval](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/eval) code, [Function](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Function/Function) code, [event handler](https://developer.mozilla.org/en-US/docs/Web/HTML/Attributes#event_handler_attributes) attributes, strings passed to [setTimeout()](https://developer.mozilla.org/en-US/docs/Web/API/setTimeout), and related functions are either function bodies or entire scripts, and invoking strict mode in them works as expected.

Similarly the old Rollup UMD factory code had issues with `this` not being `window` by default:

* https://github.com/alphagov/govuk-frontend/issues/3972#issuecomment-1699418524

To avoid any surprises by adding `<script type="module">` later rather than sooner

patrickpatrickpatrick committed:
Separate ES6 components from ES5 components

Add new config to state if application is loading ES6 components
in 'es6-components.js'. If set to true `layout_for_public` will include
a script tag with type `module` that points to `es6-components.js`.
If this config is set to true and there is no `es6-components.js` then
an error will occur.

patrickpatrickpatrick commented:
As we need to switch to Terser to compress the JS (because of govuk-frontend v5), this PR will mean you can now use ES6 in govuk_publishing_components. I would advise against it currently however as there's no target specified so it might mean you end up including JS features that don't work in all browsers that support `type="module"`.

patrickpatrickpatrick committed:
Update install-and-use docs with new JS guidance

patrickpatrickpatrick commented:
This will remain a draft until govuk-frontend v5 is released.

patrickpatrickpatrick committed:
Update govuk-frontend .js file names

In the govuk-frontend v5, the .js of UMD files has been changed to use
the "*.bundle.js" suffix. This means we have to update references to
those UMD files in the JS of the components.

patrickpatrickpatrick committed:
Remove reference to details.js

As of govuk-frontend v5, the details component does not use JS and the .js file has been removed from the package.

Remove the `data-module="govuk-details"` attribute

See 'Remove references to the JavaScript for the Details component' from https://github.com/alphagov/govuk-frontend/releases/tag/v5.0.0

patrickpatrickpatrick committed:
Update Analytics modules to use new init method

Some of the Analytics modules use window.GOVUK.Modules and so
the way they initialise needs to be updated in-line with how
other JS modules are initialised. Tests for the modules
that are effected have been updated to take into account the
new initialisation method.

patrickpatrickpatrick committed:
Update app icon paths from govuk-frontend

In govuk-frontend v5, the names of the app icons have changed and need to be updated.

In addition a manifest.json file is provided for
specifying the correct file at different sizes. These references have been updated in `layout_for_public`.

patrickpatrickpatrick committed:
Update .svg used in header component

In govuk-frontend v5, the crown .svg has been updated to also include the "GOV.UK" text. This has been updated in the header component.

Removed the `.gem-c-header__logotype` class in `_layout-header.scss` as it is not longer required

colinrotherham commented on app/views/govuk_publishing_components/components/_layout_for_public.html.erb:
```diff
@@ -79,11 +79,10 @@
-    <link rel="apple-touch-icon" href="<%= asset_path "govuk-apple-touch-icon.png" %>">
+    <link rel="mask-icon" href="<%= asset_path 'govuk-icon-mask.svg' %>" color="#0b0c0c">
+    <link rel="apple-touch-icon" sizes="180x180" href="<%= asset_path "govuk-icon-180.png" %>">
+    <link rel="apple-touch-icon" sizes="167x167" href="<%= asset_path "govuk-icon-192.png" %>">
+    <link rel="apple-touch-icon" sizes="152x152" href="<%= asset_path "govuk-icon-512.png" %>">
```
Ideally drop these lines and add `manifest.json` if you can:

```suggestion
```

<br>

The `sizes` attributes don't match up as we've replaced older Apple-specific icon sizes:

* 167×167 @1x — Home screen (iPad Pro)
* 152×152 @1x — Home screen (iPad, iPad mini)

With more [generic "maskable" icons in **manifest.json**](https://web.dev/articles/maskable-icon) with different sizes:

* 180×180
* 192×192
* 512x512

You'll see the web manifest in action when adding to home screens or the Dock

<img src="https://github.com/alphagov/govuk_publishing_components/assets/415517/ef58274e-66c7-45ba-acb9-b37a9ef423ac" alt="Add to Dock">

patrickpatrickpatrick committed:
Use .svg used in navbar

In govuk-frontend v5, the crown .svg has been updated to also include
the "GOV.UK" text. This has to be updated in the
layout_super_navigation_header component. A partial has been created
so that a variant of the .svg without "GOV.UK" can be used for the
homepage.

colinrotherham commented on app/views/govuk_publishing_components/components/layout_for_public/_header_logo.html.erb:
```diff
@@ -0,0 +1,46 @@
+      xmlns="http://www.w3.org/2000/svg"
+      viewBox="0 0 152 30"
+      height="30"
+      width="152"
```
This missed the cut off for `govuk-frontend@5.0.0-beta.2` but is an [important VoiceOver fix](https://github.com/alphagov/govuk-frontend/pull/4497)

Thanks @patrickpatrickpatrick

```suggestion
      width="152"
      aria-label="GOV.UK"
```

colinrotherham commented on app/views/govuk_publishing_components/components/layout_header/_header_logo.html.erb:
```diff
@@ -1,13 +1,17 @@
+      xmlns="http://www.w3.org/2000/svg"
+      viewBox="0 0 152 30"
+      height="30"
+      width="152"
```
Probably the [`aria-label` here too](https://github.com/alphagov/govuk_publishing_components/pull/3700/files/a80b5a65c327cfeded2e6cec43b6266e4bfcde7e..83796778727c9d77b63ff2fe39bb558242fadcdb#r1409591225)

davidtrussler commented:
I think there's another update to package file paths that is needed, to add the `dist` directory to the `app/assets/config/govuk_publishing_components_sassc-rails_manifest.js` file as below: 
```
//= link_tree ../../../node_modules/govuk-frontend/dist/govuk/assets/images
//= link_tree ../../../node_modules/govuk-frontend/dist/govuk/assets/fonts
```

MartinJJones committed:
Update GA4 modules to use new init method

Some of the GA4 modules use window.GOVUK.Modules and so the way they
initialise needs to be updated in-line with how other JS modules are
initialised. Tests for the GA4 modules that are effected have been
updated to take into account the new initialisation method.

jon-kirwan committed:
Update `lib` modules to use `init`

jon-kirwan committed:
Fix toggle specs

MartinJJones committed:
Update cross-service-header.js to use `init`

MartinJJones committed:
Update header logos

MartinJJones committed:
Ensure font-size is 19px checkboxes and warning-text

MartinJJones committed:
Remove aria-labelledby from accordion sections

MartinJJones committed:
Update feedback.js

MartinJJones commented:
Closing as we have a new PR open to upgrade to 5.1.0 - https://github.com/alphagov/govuk_publishing_components/pull/4041
